File: Semester 4/Class14_Computer_Contest_Level_3_Notes_and_Homework_Dec_10/lvl3_hw_sol/trucking_troubles.py
import sys
sys.setrecursionlimit(10000)
input = sys.stdin.readline

n, m, d = map(int, input().split())
edge = [tuple(map(int, input().split())) for i in range(m)]
p = [i for i in range(n+1)]
edge.sort(key=lambda x: x[2], reverse=True)
dest = [int(input()) for i in range(d)]

def find_set(d):   #find leader for dsu containing d
    if d != p[d]:
        p[d] = find_set(p[d])
    return p[d]

ans, i, j = 0, 0, 0
while i < d:
    while find_set(1) != find_set(dest[i]):
        fu, fv = find_set(edge[j][0]), find_set(edge[j][1])
        if fu != fv:
            p[fu] = fv
            ans = edge[j][2]
        j += 1
    i += 1
print(ans)

File: Semester 4/Class14_Computer_Contest_Level_3_Notes_and_Homework_Dec_10/lvl3_hw_sol/trucking_troubles.cpp
#include <bits/stdc++.h>
using namespace std;
const int MM = 1e4+5;
struct edge { int u, v, w;};
int n, m, d, p[MM], ans; vector<edge> a; vector<int> dest;
bool cmp(edge x, edge y){ return x.w > y.w; }
int fd(int d){ return d==p[d]? d : p[d] = fd(p[d]); }
int main(){
    ios::sync_with_stdio(0); cin.tie(0);
    cin >> n >> m >> d;
    for(int i=1; i<=n; i++) p[i] = i;
    for(int i=1, u, v, w; i<=m; i++){
        cin >> u >> v >> w; a.push_back(edge{u, v, w});
    }
    sort(a.begin(), a.end(), cmp);
    for(int i=1, x; i<=d; i++){
        cin >> x; dest.push_back(x);
    }
    for(int i=0, j=0; i<dest.size(); i++){
        for( ;fd(dest[i]) != fd(1); j++) {
            int fu = fd(a[j].u), fv = fd(a[j].v);
            if(fu != fv) { p[fu] = fv; ans = a[j].w; }
        }
    }
    cout << ans << endl;
}